contracts/exchangeIssuance/FlashMintDex.sol

```diff
@@ -0,0 +1,612 @@
+    )
+        external
+        returns (uint256)
+    {
+        uint256 totalWethReceived = 0;
+        (,, uint256[] memory wethReceived) = _getWethReceivedPerComponent(_redeemParams);
+
+        for (uint256 i = 0; i < wethReceived.length; i++) {
+            totalWethReceived += wethReceived[i];
+        }
+        return dexAdapter.getAmountOut(_swapDataWethToOutputToken, totalWethReceived);
+    }
+
+    /**
+    * Issues an exact amount of SetTokens for given amount of ETH.
+    * Any leftover ether is returned to the caller.
+    *
+    * @param _issueParams           Struct containing addresses, amounts, and swap data for issuance
+    *
+    * @return ethSpent              Amount of ETH spent
+    */
+    function issueExactSetFromETH(IssueRedeemParams memory _issueParams)
+        external
+        payable
+        isValidModuleAndSet(_issueParams.issuanceModule, address(_issueParams.setToken))
+        nonReentrant
+        returns (uint256 ethSpent)
+    {
+        require(msg.value > 0, "FlashMint: NO ETH SENT");
+
+        IWETH(WETH).deposit{value: msg.value}();
+
+        uint256 ethUsedForIssuance = _issueExactSetFromWeth(_issueParams);
+
+        require(ethUsedForIssuance <= msg.value, "FlashMint: OVERSPENT ETH");
+
+        uint256 ethReturned = msg.value.sub(ethUsedForIssuance);
+        if (ethReturned > 0) {
+            IWETH(WETH).withdraw(ethReturned);
+            payable(msg.sender).sendValue(ethReturned);
+        }
+        ethSpent = msg.value.sub(ethReturned);
+
+        emit FlashMint(msg.sender, _issueParams.setToken, IERC20(ETH_ADDRESS), ethSpent, _issueParams.amountSetToken);
+    }
+
+    /**
+    * Issues an exact amount of SetTokens for given amount of input ERC20 tokens.
+    * Any leftover value is swapped back to the payment token and returned to the caller.
+    *
+    * @param _issueParams           Struct containing addresses, amounts, and swap data for issuance
+    * @param _paymentInfo           Struct containing input token address, max amount to spend, and swap data to trade for WETH
+    *
+    * @return paymentTokenSpent     Amount of input token spent
+    */
+    function issueExactSetFromERC20(IssueRedeemParams memory _issueParams, PaymentInfo memory _paymentInfo)
+        external
+        isValidModuleAndSet(_issueParams.issuanceModule, address(_issueParams.setToken))
+        nonReentrant
+        returns (uint256 paymentTokenSpent)
+    {
+        _paymentInfo.token.safeTransferFrom(msg.sender, address(this), _paymentInfo.limitAmt);
+        uint256 wethReceived = _swapPaymentTokenForWeth(_paymentInfo.token, _paymentInfo.limitAmt, _paymentInfo.swapDataTokenToWeth);
+
+        uint256 wethSpent = _issueExactSetFromWeth(_issueParams);
+        require(wethSpent <= wethReceived, "FlashMint: OVERSPENT WETH");
+        uint256 leftoverWeth = wethReceived.sub(wethSpent);
+        uint256 paymentTokenReturned = 0;
+
+        if (leftoverWeth > 0) {
```

Comment:
There should always be some leftover WETH, so this > 0 check may not be needed.

Reply:
Added minimum refund thresholds in [7b4605b](https://github.com/IndexCoop/index-coop-smart-contracts/pull/181/commits/7b4605b146d079d3739cfdf798011ee3bc8d24b3)